File: tests/test_meta_causal_map.py
import pytest
import math
from core.engine.meta_causal_map import (
    DifferentialBoundMechanism,
    HarmonicConservationMechanism,
    SymbolicIntuitionMechanism,
    MetaCausalEngine,
    MechanismNode,
)


def test_differential_bound_mechanism_relaxation():
    bound = DifferentialBoundMechanism("bound_test", max_diff=2.0)
    bound.state[0] = 10.0
    bound.state[1] = 0.0

    initial_residual = bound.evaluate_residual()
    assert initial_residual > 0.0

    for _ in range(50):
        bound.project_structural_relaxation(learning_rate=1.0)
        bound.apply_deltas()

    final_residual = bound.evaluate_residual()
    diff = abs(bound.state[0] - bound.state[1])
    assert final_residual < 1e-3
    assert diff <= 2.01


def test_harmonic_conservation_mechanism_relaxation():
    harmonic = HarmonicConservationMechanism("harmonic_test", target_sum=15.0)
    harmonic.state[0] = 10.0
    harmonic.state[1] = 8.0
    harmonic.state[2] = 6.0  # Sum = 24.0, target = 15.0

    initial_residual = harmonic.evaluate_residual()
    assert initial_residual > 0.0

    for _ in range(20):
        harmonic.project_structural_relaxation(learning_rate=1.0)
        harmonic.apply_deltas()

    final_residual = harmonic.evaluate_residual()
    assert final_residual < 1e-3
    assert abs(sum(harmonic.state) - 15.0) < 1e-2


def test_meta_causal_engine_ecosystem():
    engine = MetaCausalEngine()

    bound = DifferentialBoundMechanism("mech_bound", max_diff=1.0)
    bound.state[0] = 5.0
    bound.state[1] = 1.0  # Diff = 4.0

    harmonic = HarmonicConservationMechanism("mech_harmonic", target_sum=10.0)
    harmonic.state[0] = 4.0
    harmonic.state[1] = 4.0
    harmonic.state[2] = 4.0  # Sum = 12.0

    engine.add_mechanism(bound)
    engine.add_mechanism(harmonic)
    engine.add_binding("mech_bound", "mech_harmonic", coupling_weight=0.1)

    initial_residual = engine.compute_total_residual()
    assert initial_residual > 0.0

    iterations = engine.step_convergence(max_iterations=100, tolerance=1e-3, learning_rate=0.5)
    final_residual = engine.compute_total_residual()

    assert iterations < 100
    assert final_residual < initial_residual

    contributions = engine.introspect_causal_contributions()
    assert "mech_bound" in contributions
    assert "mech_harmonic" in contributions


def test_symbolic_intuition_mechanism_cancellation():
    intuition = SymbolicIntuitionMechanism("concept_grounding", concept_a="Word_X", concept_b="Word_Y")
    intuition.state[0] = 7.0
    intuition.state[1] = -2.0  # Sum = 5.0 (Violation from equilibrium x + y = 0)

    initial_residual = intuition.evaluate_residual()
    assert initial_residual > 0.0

    for _ in range(20):
        intuition.project_structural_relaxation(learning_rate=1.0)
        intuition.apply_deltas()

    final_residual = intuition.evaluate_residual()
    assert final_residual < 1e-4
    assert abs(intuition.state[0] + intuition.state[1]) < 1e-3


def test_if_else_vs_constraint_relaxation():
    """
    Compares traditional if-else branching against constraint relaxation.
    In if-else, abrupt clipping causes step discontinuity.
    In constraint relaxation, equilibrium smoothly absorbs residual error.
    """
    # 1. Traditional if-else
    def traditional_if_else_control(value: float, max_limit: float) -> float:
        if value > max_limit:
            return max_limit
        return value

    clipped_val = traditional_if_else_control(10.0, 5.0)
    assert clipped_val == 5.0

    # 2. Constraint Relaxation
    bound = DifferentialBoundMechanism("smooth_bound", max_diff=5.0)
    bound.state[0] = 10.0
    bound.state[1] = 0.0  # Diff = 10.0, max = 5.0

    for _ in range(50):
        bound.project_structural_relaxation(learning_rate=1.0)
        bound.apply_deltas()

    diff = abs(bound.state[0] - bound.state[1])
    assert abs(diff - 5.0) < 0.05
    # Both states shifted smoothly to accommodate constraint
    assert bound.state[0] < 10.0
    assert bound.state[1] > 0.0

File: tests/cpp/test_meta_causal_map.cpp
#include <iostream>
#include <cassert>
#include <memory>
#include "causal_engine/core/meta_causal_map.hpp"

using namespace causal_engine;

void test_mechanism_relaxation() {
    std::cout << "\n=== Test 1: Mechanism Structural Relaxation ===\n";
    auto bound_mech = std::make_shared<DifferentialBoundMechanism>("bound_1", 2.0f);
    bound_mech->state[0] = 10.0f;
    bound_mech->state[1] = 0.0f;

    float initial_err = bound_mech->evaluate_residual();
    std::cout << "Initial bound residual: " << initial_err << " (state: " << bound_mech->state[0] << ", " << bound_mech->state[1] << ")\n";
    assert(initial_err > 0.0f);

    for (int i = 0; i < 50; ++i) {
        bound_mech->project_structural_relaxation(1.0f);
        bound_mech->apply_deltas();
    }

    float final_err = bound_mech->evaluate_residual();
    float diff = std::abs(bound_mech->state[0] - bound_mech->state[1]);
    std::cout << "Final bound residual: " << final_err << " (diff: " << diff << ", target: 2.0)\n";
    assert(final_err < 1e-3f);
    assert(diff <= 2.01f);
    std::cout << "  -> Mechanism relaxation test passed.\n";
}

void test_meta_causal_ecosystem() {
    std::cout << "\n=== Test 2: Meta-Causal Ecosystem Convergence & Introspection ===\n";
    MetaCausalEngine engine;

    auto bound = std::make_shared<DifferentialBoundMechanism>("mech_bound", 1.5f);
    bound->state[0] = 8.0f;
    bound->state[1] = 2.0f;

    auto harmonic = std::make_shared<HarmonicConservationMechanism>("mech_harmonic", 10.0f);
    harmonic->state[0] = 6.0f;
    harmonic->state[1] = 5.0f;
    harmonic->state[2] = 4.0f;

    engine.add_mechanism(bound);
    engine.add_mechanism(harmonic);

    // Add meta-causal coupling: bound residual shifts harmonic target parameter
    engine.add_binding("mech_bound", "mech_harmonic", 0.2f);

    float total_initial = engine.compute_total_residual();
    std::cout << "Initial Total Ecosystem Residual: " << total_initial << "\n";
    assert(total_initial > 0.0f);

    int steps = engine.step_convergence(100, 1e-3f, 0.5f);
    float total_final = engine.compute_total_residual();
    std::cout << "Converged in " << steps << " iterations. Final Total Residual: " << total_final << "\n";
    assert(total_final < total_initial);

    auto intro = engine.introspect_causal_contributions();
    std::cout << "Causal Contribution Introspection:\n";
    for (const auto& kv : intro) {
        std::cout << "  - Mechanism [" << kv.first << "]: " << (kv.second * 100.0f) << "%\n";
    }

    std::cout << "  -> Meta-Causal ecosystem convergence test passed.\n";
}

int main() {
    std::cout << "Running Meta-Causal Map C++ Tests...\n";
    test_mechanism_relaxation();
    test_meta_causal_ecosystem();
    std::cout << "\nAll Meta-Causal C++ Tests Passed Successfully!\n";
    return 0;
}
